### DISCUSSION TITLE
colormap alternative

### DISCUSSION THREAD

**Discussion by MrBeee** [pyqtgraph/pyqtgraph#2788] [Show and tell]
### Alternative to  **_colormap_**
Whereas there are presently 70 colormaps available, accessible through a call to ```pg.colormap.listMaps()```, there are only 12 colormaps available through a context menu of the _colormap_ parameter, when included in a dict:

```
        cmap = pg.colormap.get("viridis")
        colorParams = [
            dict(name='Color Settings', type='group', children=[
                dict(name='Bin area color',     type='color',   value=QColor(self.binAreaColor)),
                dict(name='Cmp area color',     type='color',   value=QColor(self.cmpAreaColor)),
                dict(name='Rec area color',     type='color',   value=QColor(self.recAreaColor)),
                dict(name='Src area color',     type='color',   value=QColor(self.srcAreaColor)),
                dict(name='Analysis color map', type='cmap',    value=self.analysisCmap),
                dict(name='Inactive color map', type='cmap',    value=self.inActiveCmap),
                dict(name='Opacity', type='mySlider',           value = 60.0, limits=[0,100], suffix='%'),
                dict(name='ColorMap', type='colormap', value=cmap),
            ]),
        ]
        self.parameters = pg.parametertree.Parameter.create(name='Analysis Settings', type='group', children=colorParams)
```
It is nice that the tickmarks can be edited, but as raster data is ultimately loaded into QGIS, for me this is not a requirement (detailed editing can be done in that application).

It may however be useful to select any of the 70 availabe colormaps. Therefore I started to build a **colormapList** parameter built on top of the **list** parameter . But whereas the list parameter is very flexible, it is also very unwieldy,  So I started again from the simple parameters up (bool, float, str, etc). The objective was to create a list where a colorbar pixmap is visible next to the colormap name.

I therefore came up with the following code. I hope you find it of use.

```

import pyqtgraph as pg
from pyqtgraph.parametertree import Parameter, registerParameterType
from pyqtgraph.parametertree.parameterTypes.basetypes import \
    WidgetParameterItem
from qgis.PyQt.QtCore import QSize, Qt
from qgis.PyQt.QtGui import QIcon, QPainter, QPixmap
from qgis.PyQt.QtWidgets import QComboBox


class CmapParameterItem(WidgetParameterItem):
    """
    Parameter type which displays a ComboBox containing a list of color maps
    """

    def __init__(self, param, depth):
        self.cmapList = pg.colormap.listMaps()
        self.colorbarWidth = 120                                      # Need to define *before* parent's init as parent 'makes' the widget
        super().__init__(param, depth)

    def makeWidget(self):
        w = QComboBox()
        w.setIconSize(QSize(self.colorbarWidth, 20))
        w.setStyleSheet('border: 0px')

        for item in self.cmapList:
            cmap = pg.colormap.get(item)                              # get the appropriate colormap
            brush = cmap.getBrush(span=(0.,float(self.colorbarWidth)), orientation='horizontal')
            pixmap = QPixmap (self.colorbarWidth, 20)                 # create a pixmap as starting point for a QIcon
            pixmap.fill(Qt.white)                                     # create white background; not really needed; all is covered
            painter = QPainter(pixmap)                                # paint the pixmap
            painter.setBrush(brush)
            painter.drawRect(pixmap.rect())
            painter.end()
            icon = QIcon(pixmap)                                      # create the required icon
            w.addItem(icon, item)                                     # add icon and string to the list

        w.sigChanged = w.currentIndexChanged
        w.value = w.currentText
        w.setValue = self.setValue                                    # the function defined below
        self.hideWidget = False
        return w

    def setValue(self, val):
        try:
            index = self.cmapList.index(val)
        except ValueError:
            index = 0
        self.widget.setCurrentIndex(index)


class CmapParameter(Parameter):
    itemClass = CmapParameterItem


registerParameterType('cmap', CmapParameter, override=True)
```

Note: at present ```self.colorbarWidth ``` is still hardwired in the code. Would be nice if this could be an _option_
Likewise, the list of parameters is taken from ```pg.colormap.listMaps()```. If this were to be an option (a list), one could reduce the number of colormaps available to the end user.

### Finally
The figure below gives an idea of how the cmap parameter appears in a settings dialog

![image](https://github.com/pyqtgraph/pyqtgraph/assets/38375483/dbe3ba9e-c56d-4e07-8cf5-88923c4f4353)

CET-L4 and CET-L1 are rendered using the **cmap** parameter
The viridis colormap is rendered using the **colormap** parameter (though the name dissapears after selection)

**Comment by j9ac9k** [comment]
Going to tag @NilsNemitz for this to make sure he sees it, given the amount of work he's put into making colors and such accessible in the library.

**Reply by j9ac9k** [reply]
Also going to tag @ntjess as he's the bulk of the recent work on parameter items and parameter trees.

**Reply by ntjess** [reply]
:+1: planning to look at these requests later this week

**Reply by pijyoi** [reply]
I see that the current colormap parameter wraps a `GradientWidget`. 

`ColormapDisplayItem`, as currently implemented in #2779 allows selection of the 70 bundled colormaps via its context menu.

Perhaps that could be used instead of `GradientWidget`.

**Reply by pijyoi** [reply]
As a proof of concept, I have augmented #2779 to implement `ColorMapDisplayWidget` replacing `GradientWidget` in `parametertree`. Selection of the colormaps from `pyqtgraph/colors/maps` is by context menu.

**Reply by NilsNemitz** [reply]
The implementation seems to be in good hands :)

I am all in favor of getting the extended colormap variety into the parameter tree! 

The only thing I might suggest would be allowing `self.cmapList`  to be populated manually from an optional parameter instead of always calling `pg.colormap.listMaps()`. I remember that call as not being particularly fast. It will also return many color maps that do not match the application, so a developer might want to thin out the list to what is appropriate. Finally, a user-supplied list *should* also enable loading fully custom maps from a custom location.

**Reply by pijyoi** [reply]
The proof of concept from #2779 as applied to ParameterTree looks like this now:

![Screenshot 2023-08-10 210452](https://github.com/pyqtgraph/pyqtgraph/assets/2657027/bdfcdf87-900f-42b0-ab7a-0d269942a280)
